File: js/net/src/ietf/solicit.ts
import { MSG_PARAM_NAMESPACE_COUNT, type Parameters, SetupOption, type SetupOptions } from "./parameters.ts";
import { type IetfVersion, Version } from "./version.ts";

/**
 * The MoQ Solicit extension (draft-lcurley-moq-solicit-00).
 *
 * moq-transport says nothing about whether an endpoint expects to be told what we have or
 * to ask for it, so a publisher that waits to be asked is silent against a relay that
 * never asks. This extension lets an endpoint declare, in its SETUP, that advertisements
 * to it must be solicited first.
 *
 * Declaring nothing means "no requirement, tell me unasked", which is what a peer that has
 * never heard of the extension implicitly says.
 *
 * The same draft closes the gap that settling the mechanism exposes: once advertisements
 * arrive as NAMESPACE messages on a stream we opened, nothing separates the ones the
 * publisher already had from the ones that showed up a moment later. The NAMESPACE_COUNT
 * Setup Option asks for a NAMESPACE_COUNT parameter on each SUBSCRIBE_NAMESPACE response
 * reporting the size of that initial set.
 *
 * The two are separate code points because a peer that implements only SOLICIT has never
 * heard of the parameter, and an unknown Message Parameter closes its session.
 *
 * @module
 * @internal
 */

/**
 * What the peer declared, if anything.
 *
 * The three states are distinct, and the difference between the last two is what makes the
 * requirement enforceable:
 *
 * - `undefined`: no option at all. The peer has never heard of the extension, so it cannot
 *   have honored ours and an unsolicited advertisement from it is expected.
 * - `false`: an explicit 0. No requirement of its own, but writing the option at all proves
 *   it implements this, so it is held to ours.
 * - `true`: advertisements to it must be solicited, and likewise held to ours.
 *
 * @internal
 */
export function solicitFromSetup(params: SetupOptions): boolean | undefined {
	const value = params.getVarint(SetupOption.Solicit);
	return value === undefined ? undefined : value !== 0n;
}

/**
 * Declare what we want: solicited advertisements, and a count on each response.
 *
 * Both are unconditional and true of every session: we send SUBSCRIBE_NAMESPACE for each
 * prefix we want, so there is nothing an unsolicited PUBLISH_NAMESPACE could tell us that
 * we will not have asked for, and every one of those asks wants a boundary.
 *
 * @internal
 */
export function solicitIntoSetup(params: SetupOptions, version: IetfVersion) {
	params.setVarint(SetupOption.Solicit, 1n);
	if (namespaceCountSupported(version)) {
		params.setVarint(SetupOption.NamespaceCount, 1n);
	}
}

/**
 * Whether this version carries the initial set on the SUBSCRIBE_NAMESPACE response stream
 * at all.
 *
 * Draft-14/15 answer with PUBLISH_NAMESPACE requests on streams of their own and have no
 * NAMESPACE message, so there is nothing on the response stream to count.
 *
 * @internal
 */
export function namespaceCountSupported(version: IetfVersion): boolean {
	return version !== Version.DRAFT_14 && version !== Version.DRAFT_15;
}

/**
 * Whether the peer asked us to report the count.
 *
 * Absent and 0 are the same request, unlike the declaration above: nothing here turns on
 * whether a peer that wants no count implements this.
 *
 * @internal
 */
export function namespaceCountFromSetup(params: SetupOptions, version: IetfVersion): boolean {
	if (!namespaceCountSupported(version)) return false;
	const value = params.getVarint(SetupOption.NamespaceCount);
	return value !== undefined && value !== 0n;
}

/**
 * The initial set size a SUBSCRIBE_NAMESPACE response reported, or `undefined` on a peer
 * that does not implement the extension. `0n` is a real answer: the prefix is empty.
 *
 * @internal
 */
export function namespaceCountFromResponse(params: Parameters): bigint | undefined {
	return params.vars.get(MSG_PARAM_NAMESPACE_COUNT);
}

/**
 * Report the initial set size on a SUBSCRIBE_NAMESPACE response.
 *
 * @internal
 */
export function namespaceCountIntoResponse(params: Parameters, count: bigint) {
	params.vars.set(MSG_PARAM_NAMESPACE_COUNT, count);
}

File: rs/moq-net/src/ietf/solicit.rs
//! The MoQ Solicit extension (draft-lcurley-moq-solicit-00).
//!
//! moq-transport carries no statement of what an endpoint intends to do with a session,
//! so neither side can tell whether the other will announce, ask, both, or neither. An
//! unsolicited PUBLISH_NAMESPACE is therefore a guess: either the only thing that will
//! ever tell the peer what we have, or noise it never wanted.
//!
//! This extension lets an endpoint declare, via the SOLICIT Setup Option, that
//! advertisements to it must be solicited first. Declaring nothing means "no requirement,
//! tell me unasked", which is what every peer that has never heard of this extension
//! implicitly says, so the default is the chatty, interoperable behavior.
//!
//! The option is a plain Setup Option, so it rides every draft we speak rather than
//! needing the unified SETUP the MoQ Cluster extension does.

use crate::coding::DecodeError;

use super::Version;

/// SOLICIT Setup Option: whether advertisements to the sender must be solicited. Even, so
/// the value is a bare varint.
pub const SOLICIT: u64 = 0x40B5A;

/// What the peer declared, if anything.
///
/// The three states are distinct, and the difference between the last two is what makes
/// the requirement enforceable:
///
/// - `None`: no option at all. The peer has never heard of the extension, so it cannot
///   have honored ours and an unsolicited advertisement from it is expected.
/// - `Some(false)`: an explicit 0. No requirement of its own, but writing the option at
///   all proves it implements this, so it is held to ours.
/// - `Some(true)`: advertisements to it must be solicited, and likewise held to ours.
pub fn from_setup(params: &super::Parameters, _version: Version) -> Result<Option<bool>, DecodeError> {
	Ok(params
		.get_varint(super::ParameterVarInt::Solicit)
		.map(|value| value != 0))
}

/// Declare that advertisements to us must be solicited.
///
/// Unconditional, and true of every session we open or accept: we send
/// SUBSCRIBE_NAMESPACE for every prefix we are allowed to discover, so there is nothing an
/// unsolicited PUBLISH_NAMESPACE could tell us that we will not have asked for. A peer
/// that honors it stops guessing which of the two we expect, which is the whole point.
pub fn into_setup(params: &mut super::Parameters, _version: Version) {
	params.set_varint(super::ParameterVarInt::Solicit, 1);
}

#[cfg(test)]
mod tests {
	use super::*;

	const VERSION: Version = Version::Draft19;

	/// What we declare is what a peer reads back, on every draft that carries Setup
	/// Options: an old peer can opt out too.
	#[test]
	fn every_version_round_trips() {
		for version in [Version::Draft14, Version::Draft15, Version::Draft16, Version::Draft19] {
			let mut params = super::super::Parameters::default();
			into_setup(&mut params, version);
			assert_eq!(from_setup(&params, version).unwrap(), Some(true));
		}
	}

	/// A peer that declared nothing wants to be told unasked, which is what keeps this
	/// backward compatible with every implementation that has never heard of it.
	#[test]
	fn absent_declares_nothing() {
		let params = super::super::Parameters::default();
		assert_eq!(from_setup(&params, VERSION).unwrap(), None);
	}

	/// An explicit 0 asks for the same treatment an absent option does, but it is not the
	/// same statement: writing the option proves the peer implements this, which is what
	/// lets us hold it to our own declaration.
	#[test]
	fn zero_is_a_declaration_not_an_absence() {
		let mut params = super::super::Parameters::default();
		params.set_varint(super::super::ParameterVarInt::Solicit, 0);

		assert_eq!(from_setup(&params, VERSION).unwrap(), Some(false));
	}

	/// A value this draft doesn't define still means "ask me first", so a later revision
	/// that says more can't be read as saying nothing.
	#[test]
	fn any_non_zero_requires_solicitation() {
		for value in [1, 2, 0x8000] {
			let mut params = super::super::Parameters::default();
			params.set_varint(super::super::ParameterVarInt::Solicit, value);

			assert_eq!(from_setup(&params, VERSION).unwrap(), Some(true), "value {value}");
		}
	}
}
